documentation/interrupt_handling.md

```diff
@@ -1,23 +1,88 @@
-Entry and exit of secure world
-==============================
-
-#Introduction
-Depending on configuration of the system secure world can be entered during
-different conditions. This document will describe only the configuration
-used by OP-TEE.
-
-Monitor vector is VBAR_EL3 in AArch64 and MVBAR in ARMv7/AArch32. State
-vector is any of:
-* VBAR_EL2 or VBAR_EL1 (secure and non-secure) for AArch64, depending on
-  configuration of hypervisor
-* HVBAR or VBAR (secure and non-secure) for ARMv7, depending on
-  configuration of hypervisor
-
-The processor is configured to use:
-* Monitor vector for FIQ exceptions while SCR_NS is set and state vector
-  when SCR_NS is cleared
-* Monitor vector for SMC exceptions
-* State vector for IRQ exceptions
+Exceptions, Interrupts and World Contexts
+=========================================
+
+This document describes how optee_os handles switches of
+world execution context based on SMC exceptions and interrupt notifications
+Interrupt notifications are IRQ/FIQ exceptions which may also imply switching
+of world execution context: normal world to secure world, or secure world to
+normal world.
+
+## Content
+
+1. [Context Switches and Exceptions](#1-context-switches-and-exceptions)
+2. [Normal World Invokes the optee_os Through a SMC](#2-normal-world-invokes-the-optee_os-through-a-smc)
+3. [Deliver a Non-Secure Interrupt to the Normal World](#3-deliver-a-non-secure-interrupt-to-the-normal-world)
+4. [Deliver a Secure Interrupt to the Secure World](#4-deliver-a-secure-interrupt-to-the-secure-world)
+5. [Trusted Thread Scheduling](#5-trusted-thread-scheduling)
+
+# 1. Context Switches and Exceptions
+
+## 1.1. Usecases of world context switch
+
+This section lists all the cases where optee_os is involved in world context
+switches. Optee_os executes in the secure world. World switch are done by the
+cores secure monitor level/mode, referred below as the Monitor.
+
+When the normal world invokes the secure world, the normal world executes
+a SMC instruction. The SMC exception is always trapped by the Monitor. If
+the related service targets the trusted OS, the Monitor switch to optee_os
+world execution. When the secure world returns to the normal world, optee_os
+executes a SMC that is caught by the Monitor which switches back to the normal
+world.
+
+When a secure interrupt is signaled by the ARM GIC, it shall reach the
+optee_os interrupt exception vector. If the secure world was executing,
+optee_os handles straight the interrupt from it exception vector. If the
+normal world was executing when the secure interrupt raised, the Monitor
+vector must handles the exception and invoke the optee_os to serve the
+interrupt.
+
+When a non-secure interrupt is signaled by the ARM GIC, it shall reach the
+normal world interrupt exception vector. If the normal world was
```

Comment:
minor: If the normal world **is** executing, it **will** handle ...

Same in the next sentence